waiter/src/waiter/scheduler.clj

```diff
@@ -536,57 +522,68 @@
-  [clock scheduler scheduler-state-chan timeout-chan service-id->service-description-fn available? http-client failed-check-threshold]
-  (log/info "Starting scheduler syncer")
+  and sends the data to the router state maintainer.
+
+  get-service->instances-fn is a function that returns
+  a map of scheduler/Service records -> map of scheduler/ServiceInstance records.
+  The nested map has the following keys: :active-instances and :failed-instances.
+  The active-instances should not be assumed to be healthy (or live).
+  The failed-instances are guaranteed to be dead.\""
+  [clock timeout-chan service-id->service-description-fn available? failed-check-threshold
+   scheduler-name get-service->instances-fn scheduler-state-chan]
+  (log/info "starting scheduler syncer")
   (let [exit-chan (async/chan 1)
-        state-query-chan (async/chan 32)]
+        state-query-chan (async/chan 32)
+        syncer-state-atom (atom {})]
     (async/go
       (try
-        (loop [{:keys [last-update-time service-id->health-check-context] :as current-state}
-               {:service-id->health-check-context {}}]
+        (loop [{:keys [service-id->health-check-context] :as current-state} @syncer-state-atom]
+          (reset! syncer-state-atom current-state)
           (when-let [next-state
                      (async/alt!
                        exit-chan
                        ([message]
-                         (log/warn "Stopping scheduler-syncer")
+                         (log/warn "stopping scheduler-syncer")
                          (when (not= :exit message)
                            (throw (ex-info "Stopping scheduler-syncer" {:time (t/now), :reason message}))))
 
                        state-query-chan
                        ([{:keys [response-chan service-id]}]
-                         (if service-id
-                           (let [scheduler-state (service-id->state scheduler service-id)
-                                 health-check-context (service-id->health-check-context service-id)]
-                             (async/>! response-chan (-> {:last-update-time last-update-time}
-                                                         (merge scheduler-state health-check-context))))
-                           (async/>! response-chan (merge current-state (state scheduler))))
+                         (->> (if service-id
+                                (retrieve-syncer-state current-state service-id)
+                                (retrieve-syncer-state current-state))
+                              (async/>! response-chan))
                          current-state)
 
                        timeout-chan
                        ([]
                          (try
                            (timers/start-stop-time!
-                             (metrics/waiter-timer "state" "scheduler-sync")
+                             (metrics/waiter-timer "state" "scheduler-sync" scheduler-name)
```

Comment:
Does `scheduler-name` belong before `scheduler-sync`?

Reply:
Yes, changed. We now use `"core" "scheduler" scheduler-name` to qualify scheduler metrics.